### ISSUE TITLE
Help Debugging Slow Decoding on OSX

### ISSUE THREAD

**Issue by aauren** [Xpra-org/xpra#3101]
I'm hoping that someone can help me understand why OSX is introducing so much latency on the decoding side. I'm trying to use the seamless mode of XPRA to run Intellij's GoLand.

You can see the decoding latency in the screenshots and log segment that I've included below. During the time that I'm actively using the application, especially during larger redraws (e.g. opening a new file, scrolling, etc.) CPU usage on my client spikes up to 100% of one core. These large redraws often causes user visible pauses of at least a second, I'm guessing maybe even more. However, even moving the cursor around within the file has noticeable latency.

I have tried many iterations of min-speed / min-quality and none of them have seemed to improve the experience. However, I would guess that they wouldn't change the experience if the problem is mainly on the client side. I use the following command to try to disable as much as I can:
```
xpra start ssh://<user>@<host> --video-encoders=nvenc --start=/snap/bin/goland --speaker=off --webcam=no --microphone=off --av-sync=no --clipboard=no
```

Information about the environment:
* Network: `Local LAN` (Both machines are connected via ethernet at 1Gbps to the same switch)
* Client OS / Version: `OS X Catalina (10.15.7)`
* Server OS / Version: `Ubuntu 20.10`
* Client XPRA Version: `4.1.3 64-bit`
* Server XPRA Version: `4.1.3-r1 64-bit`
* XPRA Mode: `Seamless`
* Is this a new or old issue / Regression? `This is the first time that I've tried to use this client with this server. I've used XPRA in the past (maybe a year or more) with different Linux clients / servers within this same network environment and have had a much better experience. I'm guessing that the issue is related to the OSX client at this point.`

Screenshots of session info:
![image](https://user-images.githubusercontent.com/1392295/116582974-7d64e300-a8db-11eb-9475-002b3146586b.png)
![image](https://user-images.githubusercontent.com/1392295/116583028-89e93b80-a8db-11eb-9c62-803ef7a49d00.png)
![image](https://user-images.githubusercontent.com/1392295/116582991-81910080-a8db-11eb-973b-f5c4d297bedb.png)
![image](https://user-images.githubusercontent.com/1392295/116583057-91a8e000-a8db-11eb-9ad1-f1efd0a0b2ce.png)
![image](https://user-images.githubusercontent.com/1392295/116583080-98375780-a8db-11eb-8c5f-da76f53e513a.png)

Excerpt from logs that show decoding latency (note this log was taken with an extra flag `--dpi=100` at the time of the run, but I believe that subsequent runs with the above flags is similar):
```
2021-04-29 09:16:59,668  SSH: '->server xpra-pipe-proxy: waiting for data'
2021-04-29 09:16:59,669  SSH: '->server xpra-pipe-proxy: writing 25 bytes'
2021-04-29 09:16:59,669  SSH: '->server xpra-pipe-proxy: written 25 bytes'
2021-04-29 09:16:59,669  SSH: '->server xpra-pipe-proxy: waiting for data'
2021-04-29 09:16:59,669  SSH: '<-server xpra-pipe-proxy: writing 24 bytes'
2021-04-29 09:16:59,670 processing packet ping
2021-04-29 09:16:59,670 check_server_echo(0) last=True, server_ok=True (last_ping_echoed_time=417062141)
2021-04-29 09:16:59,670 next_packet() packets in queues: priority=0, ordinary=1, mouse=False
2021-04-29 09:16:59,671  SSH: '<-server xpra-pipe-proxy: written 24 bytes'
2021-04-29 09:16:59,672  SSH: '<-server xpra-pipe-proxy: waiting for data'
2021-04-29 09:16:59,672  SSH: '->server xpra-pipe-proxy: writing 26 bytes'
2021-04-29 09:16:59,672  SSH: '->server xpra-pipe-proxy: written 26 bytes'
2021-04-29 09:16:59,672  SSH: '->server xpra-pipe-proxy: waiting for data'
2021-04-29 09:16:59,677 draw_widget(<Gtk.DrawingArea object at 0x10d7a5c80 (GtkDrawingArea at 0x7fc6e380d330)>, <cairo.Context object at 0x10d767770>)
2021-04-29 09:16:59,677 clip_to_backing(gtk3.CairoBacking(ImageSurface(2540, 1387) : size=(2540, 1387), render_size=(2540, 1387)), <cairo.Context object at 0x10d767770>) rectangle=(0, 0, 2540, 1387)
2021-04-29 09:16:59,677 cairo_draw: backing=<cairo.ImageSurface object at 0x10d767a90>, size=(2540, 1387), render-size=(2540, 1387), offsets=(0, 0, 0, 0), pointer_overlay=None
2021-04-29 09:16:59,872  SSH: '<-server xpra-pipe-proxy: writing 634 bytes'
2021-04-29 09:16:59,872  SSH: '<-server xpra-pipe-proxy: written 634 bytes'
2021-04-29 09:16:59,872 processing packet draw
2021-04-29 09:16:59,873 check_server_echo(0) last=True, server_ok=True (last_ping_echoed_time=417062141)
2021-04-29 09:16:59,873  SSH: '<-server xpra-pipe-proxy: waiting for data'
2021-04-29 09:16:59,873  SSH: '<-server xpra-pipe-proxy: writing 414 bytes'
2021-04-29 09:16:59,873 processing packet draw
2021-04-29 09:16:59,873 check_server_echo(0) last=True, server_ok=True (last_ping_echoed_time=417062141)
2021-04-29 09:16:59,873  SSH: '<-server xpra-pipe-proxy: written 414 bytes'
2021-04-29 09:16:59,873  SSH: '<-server xpra-pipe-proxy: waiting for data'
2021-04-29 09:17:00,078 wait(0.5000) actually waited 0.5001
2021-04-29 09:17:00,078 poll_UI_loop() last_UI_thread_time was 0.4 seconds ago (max 1), UI_blocked=False
2021-04-29 09:17:00,078 poll_UI_loop() ok, firing [<bound method OSXClipboardProxy.timer_clipboard_check of ClipboardProxyCore(CLIPBOARD)>]
2021-04-29 09:17:00,078 timer_clipboard_check() was 114, now 114 (have token: False)
2021-04-29 09:17:00,419 cairo_paint_border(<cairo.Context object at 0x10d767770>, None)
2021-04-29 09:17:00,419 UI_thread_wakeup() elapsed=341.01ms
2021-04-29 09:17:00,420 mask_to_names(<flags 0 of type Gdk.ModifierType>)=['mod2'] swap_keys=True, modmap={<flags GDK_SHIFT_MASK of type Gdk.ModifierType>: 'shift', <flags GDK_LOCK_MASK of type Gdk.ModifierType>: 'lock', <flags GDK_SUPER_MASK of type Gdk.ModifierType>: 'mod3', <flags GDK_HYPER_MASK of type Gdk.ModifierType>: 'mod4', <flags GDK_META_MASK of type Gdk.ModifierType>: 'control', <flags GDK_CONTROL_MASK of type Gdk.ModifierType>: 'mod1'}, num_lock_state=True, num_lock_modifier=mod2
2021-04-29 09:17:00,420 pointer_modifiers(<Gdk.EventMotion object at 0x10d75dd10 (void at 0x7fc6e4a8b950)>)=((2844, 1025), (1164, 979), ['mod2'], []) (x_root=2844.0, y_root=1025.0, window_offset=None)
2021-04-29 09:17:00,420 do_motion_notify_event(<Gdk.EventMotion object at 0x10d75dd10 (void at 0x7fc6e4a8b950)>) wid=2 / focus=2 / window wid=2, device=Core Pointer, pointer=(2844, 1025), relative pointer=(1164, 979), modifiers=['mod2'], buttons=[]
2021-04-29 09:17:00,420 send_mouse_position(['pointer-position', 2, [2844, 1025, 1164, 979], ['mod2'], []]) elapsed=775, delay left=-767
2021-04-29 09:17:00,420 do_send_mouse_position() position=['pointer-position', 2, [2844, 1025, 1164, 979], ['mod2'], []]
2021-04-29 09:17:00,420 next_packet() packets in queues: priority=0, ordinary=0, mouse=True
2021-04-29 09:17:00,420 gravity_adjust(1175, 722, {b'rgb_format': b'BGRX', b'lz4': 4, b'flush': 2, b'window-size': (2540, 1387), 'encoding': 'rgb24'}) window_size=(2540, 1387), size=(2540, 1387), gravity=NorthWest
2021-04-29 09:17:00,420 cairo._do_paint_rgb(RGB24, True, 1152 <class 'bytes'>, 1175, 722, 16, 18, 16, 18, 64, {b'rgb_format': 'BGRX', b'lz4': 4, b'flush': 2, b'window-size': (2540, 1387), 'encoding': 'rgb24'}) set_image_surface_data=<built-in function set_image_surface_data>, use pixbuf=False
2021-04-29 09:17:00,421 source image surface: (cairo.Format.RGB24, 16, 18, 64, cairo.Content.COLOR)
2021-04-29 09:17:00,421 cairo_paint_surface(<function CairoBackingBase.cairo_paint_surface.<locals>.set_source_surface at 0x10d77b8b0>, <cairo.ImageSurface object at 0x10d7679f0>, 1175, 722, 16, 18, 16, 18, {b'rgb_format': 'BGRX', b'lz4': 4, b'flush': 2, b'window-size': (2540, 1387), 'encoding': 'rgb24'}) backing=<cairo.ImageSurface object at 0x10d767a90>, paint box line width=0
2021-04-29 09:17:00,421 record_decode_time(True, ) wid=2, rgb24: 16x18, 757.6ms
2021-04-29 09:17:00,421 sending ack: ('damage-sequence', 134, 2, 16, 18, 757642, '')
2021-04-29 09:17:00,421 gravity_adjust(2501, 750, {b'flush': 1, b'window-size': (2540, 1387), 'encoding': 'png'}) window_size=(2540, 1387), size=(2540, 1387), gravity=NorthWest
2021-04-29 09:17:00,421 next_packet() packets in queues: priority=1, ordinary=0, mouse=False
2021-04-29 09:17:00,421 cairo._do_paint_rgb(RGB24, False, 13176 <class 'bytes'>, 2501, 750, 18, 244, 18, 244, 54, {b'flush': 1, b'window-size': (2540, 1387), 'encoding': 'png', b'rgb_format': 'RGB'}) set_image_surface_data=<built-in function set_image_surface_data>, use pixbuf=False
2021-04-29 09:17:00,421 source image surface: (cairo.Format.RGB24, 18, 244, 72, cairo.Content.COLOR)
2021-04-29 09:17:00,422 cairo_paint_surface(<function CairoBackingBase.cairo_paint_surface.<locals>.set_source_surface at 0x10d79ea60>, <cairo.ImageSurface object at 0x10d767f70>, 2501, 750, 18, 244, 18, 244, {b'flush': 1, b'window-size': (2540, 1387), 'encoding': 'png', b'rgb_format': 'RGB'}) backing=<cairo.ImageSurface object at 0x10d767a90>, paint box line width=0
2021-04-29 09:17:00,422 record_decode_time(True, ) wid=2, png: 18x244, 757.4ms
2021-04-29 09:17:00,422 sending ack: ('damage-sequence', 135, 2, 18, 244, 757496, '')
2021-04-29 09:17:00,422 gravity_adjust(2242, 24, {b'rgb_format': b'BGRX', b'lz4': 4, b'window-size': (2540, 1387), 'encoding': 'rgb24'}) window_size=(2540, 1387), size=(2540, 1387), gravity=NorthWest
2021-04-29 09:17:00,422 cairo._do_paint_rgb(RGB24, True, 3256 <class 'bytes'>, 2242, 24, 37, 22, 37, 22, 148, {b'rgb_format': 'BGRX', b'lz4': 4, b'window-size': (2540, 1387), 'encoding': 'rgb24'}) set_image_surface_data=<built-in function set_image_surface_data>, use pixbuf=False
2021-04-29 09:17:00,422 source image surface: (cairo.Format.RGB24, 37, 22, 148, cairo.Content.COLOR)
2021-04-29 09:17:00,422 cairo_paint_surface(<function CairoBackingBase.cairo_paint_surface.<locals>.set_source_surface at 0x10d79ec10>, <cairo.ImageSurface object at 0x10d767890>, 2242, 24, 37, 22, 37, 22, {b'rgb_format': 'BGRX', b'lz4': 4, b'window-size': (2540, 1387), 'encoding': 'rgb24'}) backing=<cairo.ImageSurface object at 0x10d767a90>, paint box line width=0
2021-04-29 09:17:00,422 next_packet() packets in queues: priority=1, ordinary=0, mouse=False
2021-04-29 09:17:00,422 record_decode_time(True, ) wid=2, rgb24: 37x22, 756.6ms
2021-04-29 09:17:00,422 sending ack: ('damage-sequence', 136, 2, 37, 22, 756607, '')
2021-04-29 09:17:00,423 after_draw_refresh(True, ) pending_refresh=[(1175, 722, 16, 18), (2501, 750, 18, 244), (2242, 24, 37, 22), (2501, 750, 18, 244)]
2021-04-29 09:17:00,423  SSH: '->server xpra-pipe-proxy: writing 31 bytes'
2021-04-29 09:17:00,423 next_packet() packets in queues: priority=1, ordinary=0, mouse=False
2021-04-29 09:17:00,423 gravity_adjust(2501, 750, {b'window-size': (2540, 1387), 'encoding': 'png'}) window_size=(2540, 1387), size=(2540, 1387), gravity=NorthWest
2021-04-29 09:17:00,423  SSH: '->server xpra-pipe-proxy: written 31 bytes'
2021-04-29 09:17:00,424 cairo._do_paint_rgb(RGB24, False, 13176 <class 'bytes'>, 2501, 750, 18, 244, 18, 244, 54, {b'window-size': (2540, 1387), 'encoding': 'png', b'rgb_format': 'RGB'}) set_image_surface_data=<built-in function set_image_surface_data>, use pixbuf=False
2021-04-29 09:17:00,424  SSH: '->server xpra-pipe-proxy: waiting for data'
2021-04-29 09:17:00,424 source image surface: (cairo.Format.RGB24, 18, 244, 72, cairo.Content.COLOR)
2021-04-29 09:17:00,424  SSH: '->server xpra-pipe-proxy: writing 23 bytes'
2021-04-29 09:17:00,424 cairo_paint_surface(<function CairoBackingBase.cairo_paint_surface.<locals>.set_source_surface at 0x10d79ee50>, <cairo.ImageSurface object at 0x10d767590>, 2501, 750, 18, 244, 18, 244, {b'window-size': (2540, 1387), 'encoding': 'png', b'rgb_format': 'RGB'}) backing=<cairo.ImageSurface object at 0x10d767a90>, paint box line width=0
2021-04-29 09:17:00,424  SSH: '->server xpra-pipe-proxy: written 23 bytes'
2021-04-29 09:17:00,424  SSH: '->server xpra-pipe-proxy: waiting for data'
2021-04-29 09:17:00,425 record_decode_time(True, ) wid=2, png: 18x244, 758.0ms
2021-04-29 09:17:00,425 sending ack: ('damage-sequence', 137, 2, 18, 244, 758093, '')
2021-04-29 09:17:00,425  SSH: '->server xpra-pipe-proxy: writing 25 bytes'
2021-04-29 09:17:00,425 after_draw_refresh(True, ) pending_refresh=[]
2021-04-29 09:17:00,425 next_packet() packets in queues: priority=1, ordinary=0, mouse=False
2021-04-29 09:17:00,425  SSH: '->server xpra-pipe-proxy: written 25 bytes'
2021-04-29 09:17:00,425 process_draw:     553    bytes for window   2, sequence      138,   37x22   at 2242,24   using  rgb24 encoding with options={b'rgb_format': b'BGRX', b'lz4': 3, b'flush': 1, b'window-size': (2540, 1387)}
2021-04-29 09:17:00,425  SSH: '->server xpra-pipe-proxy: waiting for data'
2021-04-29 09:17:00,426 draw_region(2242, 24, 37, 22, rgb24, 553 bytes, 148, {b'rgb_format': b'BGRX', b'lz4': 3, b'flush': 1, b'window-size': (2540, 1387)}, [<function WindowClient._do_draw.<locals>.record_d
ecode_time at 0x10d79eaf0>])
```

I have a full bug report log that I can send if you guys have a way to receive those privately. However, given the personal information contained in them I'm not super comfortable posting it publicly.

**Comment by totaam** [comment]
> `--video-encoders=nvenc`

Don't do that. It's a very bad idea and will definitely cause problems.
See https://github.com/Xpra-org/xpra/blob/master/docs/Usage/Encodings.md
If you are on a gigabit link, you should set `--speed=100` only, if anything.

Seeing that your client and server latencies are spiking to 5 seconds, it's very likely that your client is completely hung for long periods.
> `record_decode_time(True, ) wid=2, png: 18x244, 758.0ms`

That's also concerning: actually decoding a png of this size should take about 1 millisecond, the rest must be spent actually updating the screen.

My guess is that there's a problem with the graphics (OpenGL drivers perhaps?).
Try with `--opengl=no` and `--opengl=force`.
Or the OS just isn't allowing xpra to run at full speed:
> `UI_thread_wakeup() elapsed=341.01ms`

This line means that it took 340ms to schedule a test callback. (it should be near 0)
We use this to detect when MacOS is just refusing to service us and warn the server to slow down accordingly.
I am not aware of an API or mechanism for telling MacOS to let the application run at full speed.

**Comment by aauren** [comment]
@totaam Thanks so much for responding and for your help!

The thing which seems to make the largest difference is running with `--opengl=force`, when I enable this option, my decoding latency drops back down to 30 - 40 ms max. I had tried using `--opengl=yes` and experimenting with `--opengl=no` before, but neither of those seem to have an effect on my system for some reason.

With `--opengl=force` it definitely make the user interface useable and pretty responsive, even if I put it over my VPN connection where network latency becomes a much bigger factor (which is my ultimate use-case).

A couple of questions for you before we close out:
* I thought that running with `nvenc`, when possible, was the preferred way to run xpra as it utilized the video card on the server for encoding. In the Encodings doc that you link, under tuning `Best` it says: `The best possible setup is to use NVENC or another hardware encoder supported by libva: hardware encoders compress very well and do so incredibly fast.` Also my personal experience running it now and in the past is that this seems to improve performance. Can you explain a bit more about why you think it would be better for me to run without it or maybe what the risks are to enabling it?
* A problem that I've had in the past with xpra and a problem that I have now, is that often the UI seems really small, so I usually have to have two sets of Font UI settings, one for running my IDE under XPRA and another for running it without. I've tinkered in the past with setting `--dpi` and that seems to kinda do what I want, but it also seems to make the user interface more fuzzy and less clear. Is there a better way of adjusting this in xpra?

Thanks again for your time and help!

**Comment by aauren** [comment]
Awesome! Again, thanks so much for your help! Definitely appreciate it! 😄 xpra is running much better now.

**Comment by aauren** [comment]
@totaam Hey, I finally got a chance to test the beta client, but I tried both the dmg and the pkg both result in OSX saying that it is damaged and that it can't run the xpra.app file.

I haven't had this problem with the main installer, maybe something is wrong with the build? Let me know if I can give you more information.

**Comment by totaam** [comment]
@aauren  Damn, it runs fine here on 10.14 Mojave. Do you have the exact error message at hand?
Otherwise, I will try to spin up another VM to test tomorrow.

**Comment by totaam** [comment]
@aauren Just made a fresh build (4.2-r29034) and then ran it on a fresh / never used before MacOS 10.15 Catalina VM.

Please try again with the exact same version and post the full error, if any.

![verifying](https://user-images.githubusercontent.com/1049737/117331212-ba733b80-aec0-11eb-8ff1-a61982d8503f.png)
![running](https://user-images.githubusercontent.com/1049737/117331219-bd6e2c00-aec0-11eb-916f-b298edfb556f.png)

**Comment by aauren** [comment]
Ahh... I had to fix it with the `xattr -cr /Applications/Xpra.app` which I noticed is similar to the instructions on the Download page: https://github.com/Xpra-org/xpra/wiki/Download

I didn't need to do this the first time I downloaded the stable candidate, so I forgot that those instructions were there. After I ran the above command though I was able to move past the damaged message. I'm guessing that this is likely a problem on my side. Not totally sure, I try as hard as possible to not dive down into OSX too far. 😛 

Now that I was able to get it running. I can confirm that with the newest BETA that I no longer have to pass the force opengl option in order to get the client side decoding in check. Throughout my ~10 minute run, the decoder never got above 50ms which is vastly improved from what I was seeing before.

Thanks!
